dapps/marketplace/src/hoc/withFingerprint.js

```diff
@@ -11,28 +14,60 @@ const browserPropsWhitelist = [
   'userAgent'
 ]
 
-async function getFingerprintFn() {
-  return await new Promise(resolve => {
-    Fingerprint2.get({}, components => {
-      const values = components.map(component => component.value)
-      const hash = `V1-${Fingerprint2.x64hash128(values.join(''), 31)}`
-
-      // Select the browser properties to export along with the fingerprint.
-      const browserProps = {}
-      components
-        .filter(x => browserPropsWhitelist.includes(x.key))
-        .forEach(x => (browserProps[x.key] = x.value))
-      cachedFingerprintData = { fingerprint: hash, ...browserProps }
+function fetchFingerprint() {
+  return Fingerprint2.getPromise({}).then(components => {
+    const values = components.map(component => component.value)
+    const hash = `V1-${Fingerprint2.x64hash128(values.join(''), 31)}`
 
-      resolve(cachedFingerprintData)
-    })
+    // Select the browser properties to export along with the fingerprint.
+    const browserProps = {}
+    components
+      .filter(x => browserPropsWhitelist.includes(x.key))
+      .forEach(x => (browserProps[x.key] = x.value))
+    cachedFingerprintData = { fingerprint: hash, ...browserProps }
   })
 }
 
+async function getLocalStorageUUID() {
+  const uuid = localStorage.getItem(uuidKey)
+  cachedUUID = uuid ? uuid : cachedUUID
+}
+
+async function getCookieUUID() {
+  const cookie = document.cookie
+    .split(';')
+    .filter(cookie => cookie.split('=')[0] === uuidKey)[0]
+
+  cachedUUID = cookie ? cookie.split('=')[1] : cachedUUID
+}
+
+async function storeUUID(uuid) {
+  localStorage.setItem(uuidKey, uuid)
+  document.cookie = `${uuidKey}=${uuid}; expires=18 Dec 2050 12:00:00 UTC`
+}
+
+async function getFingerprintFn(walletId) {
+  await Promise.all([
+    fetchFingerprint(),
+    getLocalStorageUUID(),
+    getCookieUUID()
+  ])
+
+  cachedUUID = cachedUUID === undefined ? walletId : cachedUUID
+
+  cachedFingerprintData.uuid = cachedUUID
+  storeUUID(cachedUUID)
+  return cachedFingerprintData
+}
+
 const getFingerprint = memoize(getFingerprintFn)
 
 function withFingerprint(WrappedComponent) {
   const WithFingerprint = props => {
+    if (!props.wallet) {
```

Comment:
I wonder if we should still let the code execute but just not include the uuid in the fingerprint data ?

Reasoning being that we may use this WithFingerprint module in the future in other places where a wallet may not be accessible ?

Reply:
That is a great point, changing to your suggestion